# ETL Process for PandasLab Group 6 -- Climate Change

Since the 1958 the National Oceanic and Atmospheric Administration has been measuring CO${_2}$ levels at their Mauna Loa Observatory in Hawaii. It is the longest record of direct measurements of CO${_2}$ in the atmoshere. The data has shown a clear trend in increased atmospheric CO${_2}$ since then.

While we are aware that there are natural sources of CO${_2}$, our group would like to investigate if this overall rising trend is correlated with CO${_2}$ output from various contries collectively and individually.

The cleaning below will be done on [Nitroux Oxide emission data](https://data.worldbank.org/indicator/EN.ATM.NOXE.KT.CE?end=2015&locations=CN&start=1990&view=chart), [Methane emission data](), and [CO${_2}$ emission data](https://data.worldbank.org/indicator/EN.ATM.CO2E.KT?view=chart) accessed on July 13, 2022.

## Extract
The data can be extracted by simply clicking the CSV link in the Download section of the webpage for the appropriate data source linked above.

### Import the data to `pandas` DataFrame
You'll want to set the path to your files appropriate in the cell below

In [1]:
import pandas as pd
#Replace paths with your appropriate path; r is for rawstring in case you have subfolders in your path
N2O_data_path = r'N2O_data.csv'
Methane_data_path = r'Methane_data.csv'
CO2_data_path = r'CO2_data.csv'

## Nitrous Oxide
We'll begin with the Nitrous Oxide data (N${_2}$O) and importing the raw data from the CSV file

In [2]:
N2O_raw_data = pd.read_csv(N2O_data_path, header = 2)

In [3]:
N2O_raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 266 entries, 0 to 265
Data columns (total 67 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country Name    266 non-null    object 
 1   Country Code    266 non-null    object 
 2   Indicator Name  266 non-null    object 
 3   Indicator Code  266 non-null    object 
 4   1960            0 non-null      float64
 5   1961            0 non-null      float64
 6   1962            0 non-null      float64
 7   1963            0 non-null      float64
 8   1964            0 non-null      float64
 9   1965            0 non-null      float64
 10  1966            0 non-null      float64
 11  1967            0 non-null      float64
 12  1968            0 non-null      float64
 13  1969            0 non-null      float64
 14  1970            0 non-null      float64
 15  1971            0 non-null      float64
 16  1972            0 non-null      float64
 17  1973            0 non-null      flo

## Transform
You may notice off the bat that the `indicator column` is not entirely useful.
Let's remove the `indicator code` column with the following code:

In [4]:
N2O_raw_data.drop(columns = ['Indicator Code'], inplace=True)
N2O_raw_data

Country Name Country Code  \
0                          Aruba          ABW   
1    Africa Eastern and Southern          AFE   
2                    Afghanistan          AFG   
3     Africa Western and Central          AFW   
4                         Angola          AGO   
..                           ...          ...   
261                       Kosovo          XKX   
262                  Yemen, Rep.          YEM   
263                 South Africa          ZAF   
264                       Zambia          ZMB   
265                     Zimbabwe          ZWE   

                                        Indicator Name  1960  1961  1962  \
0    Nitrous oxide emissions (thousand metric tons ...   NaN   NaN   NaN   
1    Nitrous oxide emissions (thousand metric tons ...   NaN   NaN   NaN   
2    Nitrous oxide emissions (thousand metric tons ...   NaN   NaN   NaN   
3    Nitrous oxide emissions (thousand metric tons ...   NaN   NaN   NaN   
4    Nitrous oxide emissions (thousand metric tons ...   NaN   NaN   NaN   
..                                                 ...   ...   ...   ...   
261  Nitrous oxide emissions (thousand metric tons ...   NaN   NaN   NaN   
262  Nitrous oxide emissions (thousand metric tons ...   NaN   NaN   NaN   
263  Nitrous oxide emissions (thousand metric tons ...   NaN   NaN   NaN   
264  Nitrous oxide emissions (thousand metric tons ...   NaN   NaN   NaN   
265  Nitrous oxide emissions (thousand metric tons ...   NaN   NaN   NaN   

     1963  1964  1965  1966  ...      2013      2014      2015      2016  \
0     NaN   NaN   NaN   NaN  ...       NaN       NaN       NaN       NaN   
1     NaN   NaN   NaN   NaN  ...  259480.0  259870.0  262460.0  266510.0   
2     NaN   NaN   NaN   NaN  ...    4770.0    5090.0    4620.0    5050.0   
3     NaN   NaN   NaN   NaN  ...  189490.0  190400.0  194090.0  204420.0   
4     NaN   NaN   NaN   NaN  ...   18100.0   17560.0   18270.0   18350.0   
..    ...   ...   ...   ...  ...       ...       ...       ...       ...   
261   NaN   NaN   NaN   NaN  ...       NaN       NaN       NaN       NaN   
262   NaN   NaN   NaN   NaN  ...    3620.0    3650.0    3340.0    3350.0   
263   NaN   NaN   NaN   NaN  ...   19600.0   19870.0   18680.0   17920.0   
264   NaN   NaN   NaN   NaN  ...   15020.0   14720.0   15080.0   14940.0   
265   NaN   NaN   NaN   NaN  ...    5500.0    4910.0    5400.0    4980.0   

         2017      2018      2019  2020  2021  Unnamed: 66  
0         NaN       NaN       NaN   NaN   NaN          NaN  
1    268100.0  268290.0  267460.0   NaN   NaN          NaN  
2      5270.0    4690.0    5010.0   NaN   NaN          NaN  
3    205020.0  207550.0  209390.0   NaN   NaN          NaN  
4     18740.0   16840.0   18910.0   NaN   NaN          NaN  
..        ...       ...       ...   ...   ...          ...  
261       NaN       NaN       NaN   NaN   NaN          NaN  
262    3300.0    3080.0    3080.0   NaN   NaN          NaN  
263   19630.0   19240.0   18510.0   NaN   NaN          NaN  
264   14320.0   13990.0   14510.0   NaN   NaN          NaN  
265    5170.0    5320.0    5360.0   NaN   NaN          NaN  

[266 rows x 66 columns]

If you peaked at the GUI of the data sources, you may have noticed that the data does not begin until 1990. You'll notice in the DataFrame above, we see NaN for years 1960 - ...

Let's drop columns where all entries are NaN with the following code:

In [5]:
N2O_removed_NaN_columns = N2O_raw_data.dropna(axis='columns', how='all')
N2O_removed_NaN_columns

Country Name Country Code  \
0                          Aruba          ABW   
1    Africa Eastern and Southern          AFE   
2                    Afghanistan          AFG   
3     Africa Western and Central          AFW   
4                         Angola          AGO   
..                           ...          ...   
261                       Kosovo          XKX   
262                  Yemen, Rep.          YEM   
263                 South Africa          ZAF   
264                       Zambia          ZMB   
265                     Zimbabwe          ZWE   

                                        Indicator Name      1990      1991  \
0    Nitrous oxide emissions (thousand metric tons ...       NaN       NaN   
1    Nitrous oxide emissions (thousand metric tons ...  189480.0  189120.0   
2    Nitrous oxide emissions (thousand metric tons ...    2840.0    2930.0   
3    Nitrous oxide emissions (thousand metric tons ...   75050.0   78450.0   
4    Nitrous oxide emissions (thousand metric tons ...   15370.0   15410.0   
..                                                 ...       ...       ...   
261  Nitrous oxide emissions (thousand metric tons ...       NaN       NaN   
262  Nitrous oxide emissions (thousand metric tons ...    2250.0    2150.0   
263  Nitrous oxide emissions (thousand metric tons ...   19010.0   19120.0   
264  Nitrous oxide emissions (thousand metric tons ...   12640.0   12720.0   
265  Nitrous oxide emissions (thousand metric tons ...    5880.0    5290.0   

         1992      1993      1994      1995      1996  ...      2010  \
0         NaN       NaN       NaN       NaN       NaN  ...       NaN   
1    191600.0  191220.0  194000.0  196150.0  195960.0  ...  255340.0   
2      2890.0    2930.0    2760.0    2880.0    3120.0  ...    4580.0   
3     79640.0   81410.0  141120.0  142200.0  145410.0  ...  178570.0   
4     15460.0   15410.0   15370.0   15350.0   13810.0  ...   17850.0   
..        ...       ...       ...       ...       ...  ...       ...   
261       NaN       NaN       NaN       NaN       NaN  ...       NaN   
262    2220.0    2230.0    2230.0    2270.0    2280.0  ...    3450.0   
263   18850.0   19400.0   19280.0   19070.0   19860.0  ...   19760.0   
264   12820.0   12800.0   12570.0   12440.0   11660.0  ...   13880.0   
265    5350.0    4720.0    4890.0    4790.0    5830.0  ...    6120.0   

         2011      2012      2013      2014      2015      2016      2017  \
0         NaN       NaN       NaN       NaN       NaN       NaN       NaN   
1    259730.0  261270.0  259480.0  259870.0  262460.0  266510.0  268100.0   
2      4680.0    4720.0    4770.0    5090.0    4620.0    5050.0    5270.0   
3    182810.0  184650.0  189490.0  190400.0  194090.0  204420.0  205020.0   
4     18440.0   17520.0   18100.0   17560.0   18270.0   18350.0   18740.0   
..        ...       ...       ...       ...       ...       ...       ...   
261       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
262    3420.0    3430.0    3620.0    3650.0    3340.0    3350.0    3300.0   
263   19690.0   20090.0   19600.0   19870.0   18680.0   17920.0   19630.0   
264   13690.0   14730.0   15020.0   14720.0   15080.0   14940.0   14320.0   
265    6260.0    5800.0    5500.0    4910.0    5400.0    4980.0    5170.0   

         2018      2019  
0         NaN       NaN  
1    268290.0  267460.0  
2      4690.0    5010.0  
3    207550.0  209390.0  
4     16840.0   18910.0  
..        ...       ...  
261       NaN       NaN  
262    3080.0    3080.0  
263   19240.0   18510.0  
264   13990.0   14510.0  
265    5320.0    5360.0  

[266 rows x 33 columns]

We now have 33 columns.

You may have noticed that Kosovo (line 261) is missing many years of N2O data. Let's remove it from the dataframe and also remove any entries that have rows with less than 1/3 of the total timeline available. Since we have ~30 years worth, we can remove any rows that NaN in 10 of the years or more
We'll do so with the following code:

In [6]:
N2O_removed_NaN_rowsNcolumns = N2O_removed_NaN_columns.dropna(axis='rows', thresh=10)
N2O_removed_NaN_rowsNcolumns

Country Name Country Code  \
1    Africa Eastern and Southern          AFE   
2                    Afghanistan          AFG   
3     Africa Western and Central          AFW   
4                         Angola          AGO   
5                        Albania          ALB   
..                           ...          ...   
260                        Samoa          WSM   
262                  Yemen, Rep.          YEM   
263                 South Africa          ZAF   
264                       Zambia          ZMB   
265                     Zimbabwe          ZWE   

                                        Indicator Name      1990      1991  \
1    Nitrous oxide emissions (thousand metric tons ...  189480.0  189120.0   
2    Nitrous oxide emissions (thousand metric tons ...    2840.0    2930.0   
3    Nitrous oxide emissions (thousand metric tons ...   75050.0   78450.0   
4    Nitrous oxide emissions (thousand metric tons ...   15370.0   15410.0   
5    Nitrous oxide emissions (thousand metric tons ...    1560.0    1190.0   
..                                                 ...       ...       ...   
260  Nitrous oxide emissions (thousand metric tons ...      40.0      40.0   
262  Nitrous oxide emissions (thousand metric tons ...    2250.0    2150.0   
263  Nitrous oxide emissions (thousand metric tons ...   19010.0   19120.0   
264  Nitrous oxide emissions (thousand metric tons ...   12640.0   12720.0   
265  Nitrous oxide emissions (thousand metric tons ...    5880.0    5290.0   

         1992      1993      1994      1995      1996  ...      2010  \
1    191600.0  191220.0  194000.0  196150.0  195960.0  ...  255340.0   
2      2890.0    2930.0    2760.0    2880.0    3120.0  ...    4580.0   
3     79640.0   81410.0  141120.0  142200.0  145410.0  ...  178570.0   
4     15460.0   15410.0   15370.0   15350.0   13810.0  ...   17850.0   
5      1100.0    1150.0    1360.0    1340.0    1190.0  ...    1110.0   
..        ...       ...       ...       ...       ...  ...       ...   
260      30.0      40.0      40.0      40.0      40.0  ...      50.0   
262    2220.0    2230.0    2230.0    2270.0    2280.0  ...    3450.0   
263   18850.0   19400.0   19280.0   19070.0   19860.0  ...   19760.0   
264   12820.0   12800.0   12570.0   12440.0   11660.0  ...   13880.0   
265    5350.0    4720.0    4890.0    4790.0    5830.0  ...    6120.0   

         2011      2012      2013      2014      2015      2016      2017  \
1    259730.0  261270.0  259480.0  259870.0  262460.0  266510.0  268100.0   
2      4680.0    4720.0    4770.0    5090.0    4620.0    5050.0    5270.0   
3    182810.0  184650.0  189490.0  190400.0  194090.0  204420.0  205020.0   
4     18440.0   17520.0   18100.0   17560.0   18270.0   18350.0   18740.0   
5      1150.0    1150.0    1120.0    1140.0    1190.0    1220.0    1190.0   
..        ...       ...       ...       ...       ...       ...       ...   
260      50.0      50.0      50.0      60.0      60.0      60.0      60.0   
262    3420.0    3430.0    3620.0    3650.0    3340.0    3350.0    3300.0   
263   19690.0   20090.0   19600.0   19870.0   18680.0   17920.0   19630.0   
264   13690.0   14730.0   15020.0   14720.0   15080.0   14940.0   14320.0   
265    6260.0    5800.0    5500.0    4910.0    5400.0    4980.0    5170.0   

         2018      2019  
1    268290.0  267460.0  
2      4690.0    5010.0  
3    207550.0  209390.0  
4     16840.0   18910.0  
5      1080.0    1100.0  
..        ...       ...  
260      60.0      60.0  
262    3080.0    3080.0  
263   19240.0   18510.0  
264   13990.0   14510.0  
265    5320.0    5360.0  

[238 rows x 33 columns]

We now have 238 rows (i.e. 238 countries with N2O data for 20+ years)

Let's look at some summary data:

In [7]:
N2O_removed_NaN_rowsNcolumns.describe()

1990          1991          1992          1993          1994  \
count  2.380000e+02  2.380000e+02  2.380000e+02  2.380000e+02  2.380000e+02   
mean   9.539332e+04  9.495471e+04  9.467975e+04  9.396916e+04  9.883412e+04   
std    2.665102e+05  2.652942e+05  2.643013e+05  2.619061e+05  2.724360e+05   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    9.650000e+02  9.950000e+02  9.150000e+02  8.950000e+02  9.500000e+02   
50%    5.255000e+03  5.215000e+03  5.025000e+03  5.015000e+03  5.185000e+03   
75%    2.741250e+04  2.684250e+04  2.708500e+04  2.748000e+04  2.901250e+04   
max    2.278720e+06  2.265720e+06  2.258150e+06  2.241800e+06  2.332540e+06   

               1995          1996          1997          1998          1999  \
count  2.380000e+02  2.380000e+02  2.380000e+02  2.380000e+02  2.380000e+02   
mean   1.014074e+05  1.032568e+05  1.021787e+05  1.033730e+05  1.037633e+05   
std    2.806010e+05  2.861601e+05  2.821787e+05  2.853973e+05  2.872765e+05   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    9.100000e+02  8.200000e+02  8.625000e+02  8.350000e+02  8.525000e+02   
50%    4.815000e+03  5.440000e+03  5.390000e+03  5.435000e+03  5.490000e+03   
75%    2.792500e+04  2.859500e+04  2.889000e+04  2.838500e+04  2.895750e+04   
max    2.388080e+06  2.431850e+06  2.402500e+06  2.415640e+06  2.420660e+06   

       ...          2010          2011          2012          2013  \
count  ...  2.380000e+02  2.380000e+02  2.380000e+02  2.380000e+02   
mean   ...  1.204974e+05  1.237243e+05  1.246317e+05  1.251461e+05   
std    ...  3.386657e+05  3.477366e+05  3.504498e+05  3.524583e+05   
min    ...  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    ...  1.132500e+03  1.252500e+03  1.365000e+03  1.292500e+03   
50%    ...  5.460000e+03  5.735000e+03  5.315000e+03  5.310000e+03   
75%    ...  3.611500e+04  3.446000e+04  3.563750e+04  3.599250e+04   
max    ...  2.722640e+06  2.817790e+06  2.833410e+06  2.824930e+06   

               2014          2015          2016          2017          2018  \
count  2.380000e+02  2.380000e+02  2.380000e+02  2.380000e+02  2.380000e+02   
mean   1.267881e+05  1.280245e+05  1.299105e+05  1.322956e+05  1.322542e+05   
std    3.573619e+05  3.611225e+05  3.661508e+05  3.731202e+05  3.727713e+05   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    1.260000e+03  1.332500e+03  1.337500e+03  1.302500e+03  1.277500e+03   
50%    5.215000e+03  5.385000e+03  5.415000e+03  5.510000e+03  5.340000e+03   
75%    3.587500e+04  3.596250e+04  3.620500e+04  3.579750e+04  3.572250e+04   
max    2.867550e+06  2.890440e+06  2.920660e+06  2.982090e+06  2.979020e+06   

               2019  
count  2.380000e+02  
mean   1.329850e+05  
std    3.753982e+05  
min    0.000000e+00  
25%    1.302500e+03  
50%    5.645000e+03  
75%    3.665750e+04  
max    2.985720e+06  

[8 rows x 30 columns]

Let's narrow down the range of years to 2010-2015


In [8]:
years = ['Country Name', '2010', '2011', '2012', '2013', '2014', '2015']
N2O_cleaned = N2O_removed_NaN_rowsNcolumns[years]
N2O_cleaned

Country Name      2010      2011      2012      2013  \
1    Africa Eastern and Southern  255340.0  259730.0  261270.0  259480.0   
2                    Afghanistan    4580.0    4680.0    4720.0    4770.0   
3     Africa Western and Central  178570.0  182810.0  184650.0  189490.0   
4                         Angola   17850.0   18440.0   17520.0   18100.0   
5                        Albania    1110.0    1150.0    1150.0    1120.0   
..                           ...       ...       ...       ...       ...   
260                        Samoa      50.0      50.0      50.0      50.0   
262                  Yemen, Rep.    3450.0    3420.0    3430.0    3620.0   
263                 South Africa   19760.0   19690.0   20090.0   19600.0   
264                       Zambia   13880.0   13690.0   14730.0   15020.0   
265                     Zimbabwe    6120.0    6260.0    5800.0    5500.0   

         2014      2015  
1    259870.0  262460.0  
2      5090.0    4620.0  
3    190400.0  194090.0  
4     17560.0   18270.0  
5      1140.0    1190.0  
..        ...       ...  
260      60.0      60.0  
262    3650.0    3340.0  
263   19870.0   18680.0  
264   14720.0   15080.0  
265    4910.0    5400.0  

[238 rows x 7 columns]

## Methane
Let's look at the Methane data and import the raw data from the CSV file

In [9]:
Methane_raw_data = pd.read_csv(Methane_data_path, header = 2)
Methane_raw_data

Country Name Country Code  \
0                          Aruba          ABW   
1    Africa Eastern and Southern          AFE   
2                    Afghanistan          AFG   
3     Africa Western and Central          AFW   
4                         Angola          AGO   
..                           ...          ...   
261                       Kosovo          XKX   
262                  Yemen, Rep.          YEM   
263                 South Africa          ZAF   
264                       Zambia          ZMB   
265                     Zimbabwe          ZWE   

                               Indicator Name     Indicator Code  1960  1961  \
0    Methane emissions (kt of CO2 equivalent)  EN.ATM.METH.KT.CE   NaN   NaN   
1    Methane emissions (kt of CO2 equivalent)  EN.ATM.METH.KT.CE   NaN   NaN   
2    Methane emissions (kt of CO2 equivalent)  EN.ATM.METH.KT.CE   NaN   NaN   
3    Methane emissions (kt of CO2 equivalent)  EN.ATM.METH.KT.CE   NaN   NaN   
4    Methane emissions (kt of CO2 equivalent)  EN.ATM.METH.KT.CE   NaN   NaN   
..                                        ...                ...   ...   ...   
261  Methane emissions (kt of CO2 equivalent)  EN.ATM.METH.KT.CE   NaN   NaN   
262  Methane emissions (kt of CO2 equivalent)  EN.ATM.METH.KT.CE   NaN   NaN   
263  Methane emissions (kt of CO2 equivalent)  EN.ATM.METH.KT.CE   NaN   NaN   
264  Methane emissions (kt of CO2 equivalent)  EN.ATM.METH.KT.CE   NaN   NaN   
265  Methane emissions (kt of CO2 equivalent)  EN.ATM.METH.KT.CE   NaN   NaN   

     1962  1963  1964  1965  ...           2013           2014           2015  \
0     NaN   NaN   NaN   NaN  ...            NaN            NaN            NaN   
1     NaN   NaN   NaN   NaN  ...  496859.996045  501330.001080  509910.000766   
2     NaN   NaN   NaN   NaN  ...   15819.999695   16299.999237   15960.000038   
3     NaN   NaN   NaN   NaN  ...  315619.999558  323400.005162  330440.002084   
4     NaN   NaN   NaN   NaN  ...   35759.998322   34900.001526   37110.000610   
..    ...   ...   ...   ...  ...            ...            ...            ...   
261   NaN   NaN   NaN   NaN  ...            NaN            NaN            NaN   
262   NaN   NaN   NaN   NaN  ...   22360.000610   20950.000763   12260.000229   
263   NaN   NaN   NaN   NaN  ...   79029.998779   79919.998169   79160.003662   
264   NaN   NaN   NaN   NaN  ...   15729.999542   15590.000153   15819.999695   
265   NaN   NaN   NaN   NaN  ...   12340.000153   11369.999886   12260.000229   

              2016           2017           2018           2019  2020  2021  \
0              NaN            NaN            NaN            NaN   NaN   NaN   
1    515159.996608  518240.004519  519999.998191  525299.997816   NaN   NaN   
2     15829.999924   15609.999657   15909.999847   16370.000839   NaN   NaN   
3    343199.999519  349769.998014  359360.009015  366629.998177   NaN   NaN   
4     37430.000305   37619.998932   34000.000000   35090.000153   NaN   NaN   
..             ...            ...            ...            ...   ...   ...   
261            NaN            NaN            NaN            NaN   NaN   NaN   
262    9350.000381   10140.000343   10619.999886   10609.999657   NaN   NaN   
263   78739.997864   80110.000610   79839.996338   80129.997253   NaN   NaN   
264   15640.000343   15020.000458   15250.000000   15619.999886   NaN   NaN   
265   11869.999886   12060.000420   12340.000153   12550.000191   NaN   NaN   

     Unnamed: 66  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
..           ...  
261          NaN  
262          NaN  
263          NaN  
264          NaN  
265          NaN  

[266 rows x 67 columns]

## Transform
You may notice off the bat that the `indicator column` is not entirely useful.
Let's remove the `indicator code` column with the following code:

In [10]:
Methane_raw_data.drop(columns = ['Indicator Code'], inplace=True)
Methane_raw_data

Country Name Country Code  \
0                          Aruba          ABW   
1    Africa Eastern and Southern          AFE   
2                    Afghanistan          AFG   
3     Africa Western and Central          AFW   
4                         Angola          AGO   
..                           ...          ...   
261                       Kosovo          XKX   
262                  Yemen, Rep.          YEM   
263                 South Africa          ZAF   
264                       Zambia          ZMB   
265                     Zimbabwe          ZWE   

                               Indicator Name  1960  1961  1962  1963  1964  \
0    Methane emissions (kt of CO2 equivalent)   NaN   NaN   NaN   NaN   NaN   
1    Methane emissions (kt of CO2 equivalent)   NaN   NaN   NaN   NaN   NaN   
2    Methane emissions (kt of CO2 equivalent)   NaN   NaN   NaN   NaN   NaN   
3    Methane emissions (kt of CO2 equivalent)   NaN   NaN   NaN   NaN   NaN   
4    Methane emissions (kt of CO2 equivalent)   NaN   NaN   NaN   NaN   NaN   
..                                        ...   ...   ...   ...   ...   ...   
261  Methane emissions (kt of CO2 equivalent)   NaN   NaN   NaN   NaN   NaN   
262  Methane emissions (kt of CO2 equivalent)   NaN   NaN   NaN   NaN   NaN   
263  Methane emissions (kt of CO2 equivalent)   NaN   NaN   NaN   NaN   NaN   
264  Methane emissions (kt of CO2 equivalent)   NaN   NaN   NaN   NaN   NaN   
265  Methane emissions (kt of CO2 equivalent)   NaN   NaN   NaN   NaN   NaN   

     1965  1966  ...           2013           2014           2015  \
0     NaN   NaN  ...            NaN            NaN            NaN   
1     NaN   NaN  ...  496859.996045  501330.001080  509910.000766   
2     NaN   NaN  ...   15819.999695   16299.999237   15960.000038   
3     NaN   NaN  ...  315619.999558  323400.005162  330440.002084   
4     NaN   NaN  ...   35759.998322   34900.001526   37110.000610   
..    ...   ...  ...            ...            ...            ...   
261   NaN   NaN  ...            NaN            NaN            NaN   
262   NaN   NaN  ...   22360.000610   20950.000763   12260.000229   
263   NaN   NaN  ...   79029.998779   79919.998169   79160.003662   
264   NaN   NaN  ...   15729.999542   15590.000153   15819.999695   
265   NaN   NaN  ...   12340.000153   11369.999886   12260.000229   

              2016           2017           2018           2019  2020  2021  \
0              NaN            NaN            NaN            NaN   NaN   NaN   
1    515159.996608  518240.004519  519999.998191  525299.997816   NaN   NaN   
2     15829.999924   15609.999657   15909.999847   16370.000839   NaN   NaN   
3    343199.999519  349769.998014  359360.009015  366629.998177   NaN   NaN   
4     37430.000305   37619.998932   34000.000000   35090.000153   NaN   NaN   
..             ...            ...            ...            ...   ...   ...   
261            NaN            NaN            NaN            NaN   NaN   NaN   
262    9350.000381   10140.000343   10619.999886   10609.999657   NaN   NaN   
263   78739.997864   80110.000610   79839.996338   80129.997253   NaN   NaN   
264   15640.000343   15020.000458   15250.000000   15619.999886   NaN   NaN   
265   11869.999886   12060.000420   12340.000153   12550.000191   NaN   NaN   

     Unnamed: 66  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
..           ...  
261          NaN  
262          NaN  
263          NaN  
264          NaN  
265          NaN  

[266 rows x 66 columns]

If you peaked at the GUI of the data sources, you may have noticed that the data does not begin until 1990. You'll notice in the DataFrame above, we see NaN for years 1960 - ...

Let's drop columns where all entries are NaN with the following code:

In [11]:
Methane_removed_NaN_columns = Methane_raw_data.dropna(axis='columns', how='all')
Methane_removed_NaN_columns

Country Name Country Code  \
0                          Aruba          ABW   
1    Africa Eastern and Southern          AFE   
2                    Afghanistan          AFG   
3     Africa Western and Central          AFW   
4                         Angola          AGO   
..                           ...          ...   
261                       Kosovo          XKX   
262                  Yemen, Rep.          YEM   
263                 South Africa          ZAF   
264                       Zambia          ZMB   
265                     Zimbabwe          ZWE   

                               Indicator Name           1990           1991  \
0    Methane emissions (kt of CO2 equivalent)            NaN            NaN   
1    Methane emissions (kt of CO2 equivalent)  320589.996671  321170.003211   
2    Methane emissions (kt of CO2 equivalent)    6739.999771    7019.999981   
3    Methane emissions (kt of CO2 equivalent)  225169.999979  244579.997182   
4    Methane emissions (kt of CO2 equivalent)   21219.999313   21489.999771   
..                                        ...            ...            ...   
261  Methane emissions (kt of CO2 equivalent)            NaN            NaN   
262  Methane emissions (kt of CO2 equivalent)    8890.000343    8899.999619   
263  Methane emissions (kt of CO2 equivalent)   63189.998627   64120.002747   
264  Methane emissions (kt of CO2 equivalent)   12029.999733   12189.999580   
265  Methane emissions (kt of CO2 equivalent)    9609.999657    8720.000267   

              1992           1993           1994           1995  \
0              NaN            NaN            NaN            NaN   
1    326880.001700  330319.996810  339339.997625  347470.001196   
2      7130.000114    7210.000038    7469.999790    7829.999924   
3    253969.999358  259030.004174  264790.000498  263709.999636   
4     21950.000763   22579.999924   23479.999542   25229.999542   
..             ...            ...            ...            ...   
261            NaN            NaN            NaN            NaN   
262    8750.000000   10659.999847   14729.999542   14659.999847   
263   63750.000000   65739.997864   66739.997864   67300.003052   
264   12270.000458   12180.000305   12060.000420   11979.999542   
265    9329.999924    7800.000191    8470.000267    8869.999886   

              1996  ...           2010           2011           2012  \
0              NaN  ...            NaN            NaN            NaN   
1    351490.002086  ...  503890.006164  508730.002934  494579.996178   
2      8670.000076  ...   15199.999809   15500.000000   15630.000114   
3    269409.994841  ...  300400.002398  308019.998342  310789.995939   
4     25389.999390  ...   36900.001526   36639.999390   35770.000458   
..             ...  ...            ...            ...            ...   
261            NaN  ...            NaN            NaN            NaN   
262   15029.999733  ...   23049.999237   22329.999924   19799.999237   
263   66900.001526  ...   77260.002136   77260.002136   79209.999084   
264   11250.000000  ...   14409.999847   13869.999886   15460.000038   
265   10340.000153  ...   12270.000458   12640.000343   12460.000038   

              2013           2014           2015           2016  \
0              NaN            NaN            NaN            NaN   
1    496859.996045  501330.001080  509910.000766  515159.996608   
2     15819.999695   16299.999237   15960.000038   15829.999924   
3    315619.999558  323400.005162  330440.002084  343199.999519   
4     35759.998322   34900.001526   37110.000610   37430.000305   
..             ...            ...            ...            ...   
261            NaN            NaN            NaN            NaN   
262   22360.000610   20950.000763   12260.000229    9350.000381   
263   79029.998779   79919.998169   79160.003662   78739.997864   
264   15729.999542   15590.000153   15819.999695   15640.000343   
265   12340.000153   11369.999886   12260.000229   11869.999886   

       

We now have 33 columns.

You may have noticed that Kosovo (line 261) is missing many years of methane data. Let's remove it from the dataframe and also remove any entries that have rows with less than 1/3 of the total timeline available. Since we have ~30 years worth, we can remove any rows that NaN in 10 of the years or more
We'll do so with the following code:

In [12]:
Methane_removed_NaN_rowsNcolumns = Methane_removed_NaN_columns.dropna(axis='rows', thresh=10)
Methane_removed_NaN_rowsNcolumns

Country Name Country Code  \
1    Africa Eastern and Southern          AFE   
2                    Afghanistan          AFG   
3     Africa Western and Central          AFW   
4                         Angola          AGO   
5                        Albania          ALB   
..                           ...          ...   
260                        Samoa          WSM   
262                  Yemen, Rep.          YEM   
263                 South Africa          ZAF   
264                       Zambia          ZMB   
265                     Zimbabwe          ZWE   

                               Indicator Name           1990           1991  \
1    Methane emissions (kt of CO2 equivalent)  320589.996671  321170.003211   
2    Methane emissions (kt of CO2 equivalent)    6739.999771    7019.999981   
3    Methane emissions (kt of CO2 equivalent)  225169.999979  244579.997182   
4    Methane emissions (kt of CO2 equivalent)   21219.999313   21489.999771   
5    Methane emissions (kt of CO2 equivalent)    3539.999962    3500.000000   
..                                        ...            ...            ...   
260  Methane emissions (kt of CO2 equivalent)     209.999993     209.999993   
262  Methane emissions (kt of CO2 equivalent)    8890.000343    8899.999619   
263  Methane emissions (kt of CO2 equivalent)   63189.998627   64120.002747   
264  Methane emissions (kt of CO2 equivalent)   12029.999733   12189.999580   
265  Methane emissions (kt of CO2 equivalent)    9609.999657    8720.000267   

              1992           1993           1994           1995  \
1    326880.001700  330319.996810  339339.997625  347470.001196   
2      7130.000114    7210.000038    7469.999790    7829.999924   
3    253969.999358  259030.004174  264790.000498  263709.999636   
4     21950.000763   22579.999924   23479.999542   25229.999542   
5      3410.000086    3490.000010    4079.999924    4079.999924   
..             ...            ...            ...            ...   
260     200.000003     209.999993     239.999995     239.999995   
262    8750.000000   10659.999847   14729.999542   14659.999847   
263   63750.000000   65739.997864   66739.997864   67300.003052   
264   12270.000458   12180.000305   12060.000420   11979.999542   
265    9329.999924    7800.000191    8470.000267    8869.999886   

              1996  ...           2010           2011           2012  \
1    351490.002086  ...  503890.006164  508730.002934  494579.996178   
2      8670.000076  ...   15199.999809   15500.000000   15630.000114   
3    269409.994841  ...  300400.002398  308019.998342  310789.995939   
4     25389.999390  ...   36900.001526   36639.999390   35770.000458   
5      3910.000086  ...    3029.999971    3019.999981    3059.999943   
..             ...  ...            ...            ...            ...   
260     239.999995  ...     270.000011     280.000001     280.000001   
262   15029.999733  ...   23049.999237   22329.999924   19799.999237   
263   66900.001526  ...   77260.002136   77260.002136   79209.999084   
264   11250.000000  ...   14409.999847   13869.999886   15460.000038   
265   10340.000153  ...   12270.000458   12640.000343   12460.000038   

              2013           2014           2015           2016  \
1    496859.996045  501330.001080  509910.000766  515159.996608   
2     15819.999695   16299.999237   15960.000038   15829.999924   
3    315619.999558  323400.005162  330440.002084  343199.999519   
4     35759.998322   34900.001526   37110.000610   37430.000305   
5      3059.999943    3089.999914    3109.999895    3089.999914   
..             ...            ...            ...            ...   
260     280.000001     289.999992     300.000012     310.000002   
262   22360.000610   20950.000763   12260.000229    9350.000381   
263   79029.998779   79919.998169   79160.003662   78739.997864   
264   15729.999542   15590.000153   15819.999695   15640.000343   
265   12340.000153   11369.999886   12260.000229   11869.999886   

       

We now have 238 rows (i.e. 239 countries with Methane data for 20+ years)

Let's narrow down the year range from 2010-2015

In [13]:
Methane_cleaned = Methane_removed_NaN_rowsNcolumns[years]
Methane_cleaned

Country Name           2010           2011           2012  \
1    Africa Eastern and Southern  503890.006164  508730.002934  494579.996178   
2                    Afghanistan   15199.999809   15500.000000   15630.000114   
3     Africa Western and Central  300400.002398  308019.998342  310789.995939   
4                         Angola   36900.001526   36639.999390   35770.000458   
5                        Albania    3029.999971    3019.999981    3059.999943   
..                           ...            ...            ...            ...   
260                        Samoa     270.000011     280.000001     280.000001   
262                  Yemen, Rep.   23049.999237   22329.999924   19799.999237   
263                 South Africa   77260.002136   77260.002136   79209.999084   
264                       Zambia   14409.999847   13869.999886   15460.000038   
265                     Zimbabwe   12270.000458   12640.000343   12460.000038   

              2013           2014           2015  
1    496859.996045  501330.001080  509910.000766  
2     15819.999695   16299.999237   15960.000038  
3    315619.999558  323400.005162  330440.002084  
4     35759.998322   34900.001526   37110.000610  
5      3059.999943    3089.999914    3109.999895  
..             ...            ...            ...  
260     280.000001     289.999992     300.000012  
262   22360.000610   20950.000763   12260.000229  
263   79029.998779   79919.998169   79160.003662  
264   15729.999542   15590.000153   15819.999695  
265   12340.000153   11369.999886   12260.000229  

[239 rows x 7 columns]